### ISSUE TITLE
Drasil Research Group Meeting, 3pm, Monday July 15, 2024, ITB/225

### ISSUE THREAD

**Issue by smiths** [JacquesCarette/Drasil#3842]
0. Introduction of Xinlu 
1. #3846 (@NoahCardoso)
2. #3821 (tentative, since @JacquesCarette will be absent) (@B-rando1)
3. Deciding on which prototype to proceed with (https://github.com/JacquesCarette/Drasil/discussions/3835#discussioncomment-10013926). (@BilalM04)
4. Discussion on the structure of the Makefile (https://github.com/JacquesCarette/Drasil/pull/3823#issuecomment-2224015715). (@BilalM04)
5. Status of previous work on testing projectile motion?  Verification and validation plan? (@samm82)
6. Discuss TT candidates in future meetings

Regrets: @JacquesCarette

**Comment by smiths** [comment]
@NoahCardoso, @B-rando1 and @BilalM04, please add your weekly summaries at the end of this week.

**Comment by B-rando1** [comment]
### Weekly Summary
- I spent the most time separating GOOL into its OO, Procedural, and Common components, as was logged in this [thread](https://github.com/JacquesCarette/Drasil/discussions/3792#discussioncomment-9962372).  As of now the interface is fully separated.  Next steps are to split `drasil-gool` into different packages to avoid name conflicts, convert the Julia renderer to the new set of typeclasses, and enable `drasil-code` to use the new typeclasses.  This will still be a lot of work, but progress is going well.
- I spent some time updating the Julia renderer to only use functionality that will be in v1 of the generic procedural language.  There were a few implementations I had to change to avoid using OO functionality, but on the whole it was simple enough.
- I spent some more time on #3821.  I successfully added a `Scope` (i.e. `global` vs `local`) parameter to variables and constants, but then I got a bit stuck, so I haven't worked on it in the last few days.  The problem is that different languages have different ideas about what scopes are `global` vs `local`.  For example, in Python, Swift, and Julia, the `main` function is a script that exists in the `global` scope, while in Java, C#, and C++, the `main` function is a function that has its own local scope.  There may be other peculiarities down the road too, it's hard to say.  Our current idea is to be a bit more granular than `local` and `global`, and instead give information about the type of body the variable is tied to, e.g. `global`, `main`, `function`, `loop`, etc., and let each renderer decide what to do with that information.  I think this will be enough, but I'm still wrapping my head around what will all be involved.

### Blockers/Agenda
I'd say that #3821 is what I'm most stuck on.  I'd like to have a better plan for where I'm going with this, as I really don't know how we'll actually use the new part of the API to add `global` keywords in the Julia renderer.  This might be difficult without @JacquesCarette though, so it might not be worth discussing the meeting.  I'll think about it more today, and see if we can get a bit further.  The good news is that we can keep going on Julia without this being fixed, as the situations where we need the `global` keyword are somewhat specific.  What we'll really need it for is when we get the 'straight-shot, no functions' generator option.

**Comment by BilalM04** [comment]
## Weekly Summary 

- #3823.
  - Made minor changes, addressed comments and requested changes in the PR.
- Developed prototypes for rendering assets in mdBook. Detailed in https://github.com/JacquesCarette/Drasil/discussions/3835#discussioncomment-10013926.
  - CSV approach demonstrated in #3848.
  - Makefile approach demonstrated in #3847. 
- #3664 
  - Added ShellCheck to the Linter workflow.
  - Implemented the suggestions from ShellCheck.
  - Added ShellCheck to Makefile and Wiki.
 
## Blockers / Agenda

Deciding on which prototype to proceed with (https://github.com/JacquesCarette/Drasil/discussions/3835#discussioncomment-10013926). Shouldn't be a long discussion or it can be done in the discussion post.

The assets are the last remaining item for #3722. After the PRs for that ticket is merged in, we can start hosting on the website.

**Comment by NoahCardoso** [comment]
Summary
- I have spent most of my time this week working on #2655, I have got `set`s working in `Expr` and adding new constraints. This is what I have currently. #3846. I have just started working on adding `set`s to GOOL. @B-rando1 has been helping me to better understand how to add `set`s to GOOL. 
![image](https://github.com/user-attachments/assets/3ee005c6-a6b9-43b2-81a3-4fbda8b47d5e)
 
- #3828, After Monday's meeting I looked more into ordering alphabetically by the symbol for #3761. However, I have not worked on this issue as I have spent most of my time with the above issue.
